datafusion/physical-expr-common/src/stats_v2.rs

```diff
@@ -0,0 +1,795 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::OnceLock;
+
+use crate::stats_v2::StatisticsV2::{Bernoulli, Exponential, Gaussian, Uniform, Unknown};
+
+use arrow::datatypes::DataType;
+use datafusion_common::{internal_err, Result, ScalarValue};
+use datafusion_expr_common::interval_arithmetic::Interval;
+
+static LN_TWO_LOCK: OnceLock<ScalarValue> = OnceLock::new();
+
+/// Returns a ln(2) as a [`ScalarValue`]
+fn get_ln_two() -> &'static ScalarValue {
+    LN_TWO_LOCK.get_or_init(|| ScalarValue::Float64(Some(2_f64.ln())))
+}
+
+/// New, enhanced `Statistics` definition, represents five core definitions.
+#[derive(Clone, Debug, PartialEq)]
+pub enum StatisticsV2 {
+    Uniform {
+        interval: Interval,
+    },
+    /// f(x, λ, offset) = (λe)^(-λ(x - offset)), if x >= offset
+    Exponential {
+        rate: ScalarValue,
+        offset: ScalarValue,
+    },
+    Gaussian {
+        mean: ScalarValue,
+        variance: ScalarValue,
+    },
+    /// `p` always has [`DataType::Float64`]
+    Bernoulli {
+        p: ScalarValue,
+    },
+    Unknown {
+        mean: Option<ScalarValue>,
+        median: Option<ScalarValue>,
+        variance: Option<ScalarValue>,
+        range: Interval,
+    },
+}
+
+impl StatisticsV2 {
+    /// Constructs a new [`StatisticsV2`] with [`Uniform`] distribution from given [`Interval`],
+    /// and checks newly created statistic on validness.
+    pub fn new_uniform(interval: Interval) -> Result<Self> {
+        if interval.data_type().eq(&DataType::Boolean) {
+            return internal_err!(
+                "Construction of boolean Uniform statistic is prohibited.\
+             Create Bernoulli statistic instead."
+            );
+        }
+
+        let stat = Uniform { interval };
+        if stat.is_valid()? {
+            Ok(stat)
+        } else {
+            internal_err!("Tried to construct invalid Uniform statistic")
+        }
```

Comment:
```suggestion
        Ok(Uniform { interval })
```
Seems like we don't check `Uniform` variants in the `is_valid` function. We have two options here:
1. Add a match arm to `is_valid` to ensure that neither the lower nor the upper bound of range is infinite. This would give us a strict, proper distribution.
2. Allow improper uniform distributions with unbounded endpoints. In this case, we can use the suggestion in this comment.

Let's go with option (2) for now, we will switch to (1) in the future if this proves to be a problem.

Reply:
https://github.com/synnada-ai/datafusion-upstream/pull/57#discussion_r1938308801

How do we represent infinite bounds then?